---
title: 'Genomic Intervals and R: Exercises'
author: "Altuna Akalin"
date: "22 Sep 2015"
output: 
  html_document:
          toc: true
          toc_depth: 1
          number_sections: true
---



# The setup

Set your working directory to the source file location in RStudio top menu:

'Session > Set Working Directory > To Source File Location'

or copy paste this your your R console in your VM.

```{r,eval=FALSE}
setwd("~/compgen2015/day6_GenomicIntervals/exercises")

```

if you have done it correctly you should see this script when you type dir()


```
> dir()
[1] "GenomicInterval.exercises.html"
```

The data for the exercises is located at `../data` folder.

Run the following to see the data files.
```
dir("../data")
```

# Operations on Genomic Intervals with GenomicRanges package

##
Create a GRanges object using the information in the table below:

| chr  | start | end  |strand | score | 
| :--- |:------| :-----| :-----|:-----|
| chr1 | 10000 | 10300 |  +    | 10 |
| chr1 | 11100 | 11500 |  -    | 20 |
| chr2 | 20000 | 20030 |  +    | 15 |

##
use `start()`, `end()`, `strand()`,`seqnames()` and `width()` functions on the GRanges
object you created. Figure out what they are doing. Can you get a subset of GRanges object for intervals that are only on + strand? If you can do that, try getting intervals that are on chr1. *HINT:* GRanges objects can be subset using [ ] operator similar to data.frames but you may need
to use `start()`, `end()` and `strand()`,`seqnames()` within the [ ].

```{r,eval=FALSE,echo=FALSE}
gr0=GRanges(seqnames=c("chr1","chr1","chr2"),
        range=IRanges(start=c(10000,11100,2000),
                      end=c(10300,11500,20030)),
        score=c(10,20,15))

start(gr0)
end(gr0)
strand(gr0)
seqnames(gr0)
width(gr0)

gr0[seqnames(gr0)=="chr1",]

```


##
Import mouse (mm9 assembly) CpG islands and refseq transcripts for chr12 from UCSC browser as `GRanges` objects using `rtracklayer` functions. HINT: Check the lecture material and modify the code there as necessary. If that somehow does not work, go to UCSC browser and download it as a BED file. The trackname for Refseq genes is "RefSeq Genes" and table name is "refGene". 

```{r,eval=FALSE,echo=FALSE}

require(rtracklayer)
session <- browserSession("UCSC",url='http://genome-euro.ucsc.edu/cgi-bin/')
genome(session) <- "mm9"

## choose CpG island track on chr12
cpgi <- ucscTableQuery(session, track="CpG Islands",table="cpgIslandExt",
        range=GRangesForUCSCGenome("mm9", "chr12"))

## get the GRanges object for the track
cpgi.gr=track(cpgi)

## choose CpG island track on chr12
query <- ucscTableQuery(session, track="RefSeq Genes",table="refGene",
        range=GRangesForUCSCGenome("mm9", "chr12"))
## get the GRanges object for the track
ref.gr=track(query)


```

##
Following from the exercise above, get the promoters of Refseq transcripts (-1000bp and +1000 bp of the TSS) and calculate what percentage of them overlap with CpG islands. HINT: You have to get the promoter coordinates and use `findOverlaps()` or `subsetByOverlaps()` from `GenomicRanges` package. To get promoters, type `?promoters` on the R console and see how to use  that function to get promoters or calculate their coordinates as shown in the lecture material.

```{r,eval=FALSE,echo=FALSE}
proms=promoters(ref.gr, upstream=1000, downstream=1000)
sub.cpgi=subsetByOverlaps(cpgi.gr,proms)

100*length(sub.cpgi)/length(cpgi.gr)

```


## 
Plot the distribution of CpG island lengths for CpG islands that overlap with the 
promoters.

```{r,eval=FALSE,echo=FALSE}
hist(width(sub.cpgi))
median(width(sub.cpgi))

```


##
Get canonical peaks for SP1 (peaks that are in both replicates) on chr21. Peaks for each replicate are located in `../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz` and `../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz` files. HINT: You need to use `findOverlaps()` or `subsetByOverlaps()` to get the subset of peaks that occur in both replicates. You can try to read *broadPeak.gz files using genomation function `readBroadPeak`, broadPeak is just an extended BED format.
EXTRA credit: Try use `coverage()` and slice`()` functions to get canonical peaks.

```{r,eval=FALSE}
library(genomation)
rep1=readBroadPeak("../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep1.broadPeak.gz")
rep2=readBroadPeak("../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xPkRep2.broadPeak.gz")
```

```{r,eval=FALSE,echo=FALSE}
subsetByOverlaps(rep1,rep2)
```

# Dealing with mapped high-throughput sequencing reads

## 
Count the reads overlapping with canonical Sp1 peaks using the BAM file for one of the replicates: `../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam`. **HINT**: 
Use functions from `GenomicAlignments`, see [lecture notes](../lectures/genomicIntervalsTutorial.html).

```{r,eval=FALSE,echo=FALSE}

canon=subsetByOverlaps(rep1,rep2)

library(Rsamtools)
bamfile="../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam"

# get reads for regions of interest from the bam file
param <- ScanBamParam(which=canon)
counts=countBam(bamfile, param=param)

```

# Dealing with contiguous scores over the genome

## 
Extract Views object for the promoters on chr20 from `../data/H1.ESC.H3K4me1.chr20.bw` file. Plot the first "View" as a line plot. **HINT**: see [lecture notes](../lectures/genomicIntervalsTutorial.html), adapt the code from there.

## 
Make a histogram of the maximum signal for the Views in the object you extracted above. You can use any of the view summary functions or use lapply() and write your own summary function.

## 
Get the genomic positions of maximum signal in each view and make a GRanges object. **HINT**: See ?viewRangeMaxs help page. Try to make a GRanges object out of the returned object.

# Visualizing and summarizing genomic intervals

##
Extract -500,+500 bp regions around TSSes on chr21, there are refseq files in the `../data` folder or you can
pull the data out of UCSC browser. Use SP1 ChIP-seq data 
(`../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam` ) to create an average profile of read coverage around TSSes. Following that, visualize the read coverage with a heatmap. **HINT**: All of these possible using `genomation` package functions.

```{r}
gene.chr21=readTranscriptFeatures("../data/refseq.hg19.chr21.bed",up.flank = 500,down.flank = 500)
prom.chr21=gene.chr21$promoters

sm=ScoreMatrix("../data/wgEncodeHaibTfbsGm12878Sp1Pcr1xAlnRep1.chr21.bam",prom.chr21,
               type="bam",strand.aware = TRUE)
plotMeta(sm)

```

##
Extract -500,+500 bp regions around TSSes on chr20. Use H3K4me3 (`../data/H1.ESC.H3K4me3.chr20.bw`) and H3K27ac (`../data/H1.ESC.H3K27ac.chr20.bw`) ChIP-seq enrichment data in the data folder and create heatmaps and average signal profiles for regions around the TSSes.

```{r}
gene.chr20=readTranscriptFeatures("../data/refseq.hg19.chr20.bed",up.flank = 500,down.flank = 500)
prom.chr20=gene.chr20$promoters

sml=ScoreMatrixList(c("../data/H1.ESC.H3K4me3.chr20.bw",
                      "../data/H1.ESC.H3K27ac.chr20.bw"),prom.chr20,
                      type="bigWig",strand.aware = TRUE)

# look for the average enrichment
plotMeta(sml, profile.names = c("H3K4me3","H3K27ac"), xcoords = c(-500,500),
         ylab="log2enrichment",dispersion = "se",
         xlab="bases around TSS")

multiHeatMatrix(sml,order=TRUE,winsorize = c(0,95),
                matrix.main = c("H3K4me3","H3K27ac"))

```

## 
Visualize one of the -500,+500 bp regions around TSS using `Gviz` functions. You should visualize both H3K4me3 and H3K27ac and the gene models.

```{r}
library(Gviz)
H3K4me3.track=DataTrack("../data/H1.ESC.H3K4me3.chr20.bw",type = "l",chromosome = "chr20"
                    name="H3K4me3")

H3K27ac.track=DataTrack("../data/H1.ESC.H3K27ac.chr20.bw",type = "l",
                    name="H3K27ac")

track.list=list(H3K4me3.track,H3K27ac.track)

prom.chr20[1,]
prom.chr20[2,]
prom.chr20[729,]
plotTracks(track.list,from=49547020,to=49548020,
           chromosome="chr20")



```
